src/ICS20Transfer.sol

```diff
@@ -81,16 +83,15 @@ contract ICS20Transfer is IIBCApp, IICS20Transfer, IICS20Errors, Ownable, Reentr
 
         address sender = ICS20Lib.mustHexStringToAddress(packetData.sender);
 
-        (address erc20Address, bool originatorChainIsSource) = getSendERC20AddressAndSource(msg_.packet, packetData);
-
-        // transfer the tokens to us (requires the allowance to be set)
-        _transferFrom(sender, address(this), erc20Address, packetData.amount);
+        (address erc20Address, bool senderChainIsSource) = getSendERC20AddressAndSource(msg_.packet, packetData);
 
-        if (!originatorChainIsSource) {
+        if (senderChainIsSource) {
+            // transfer the tokens to the escrow address
+            address ics20EscrowAddress = _getEscrowAddress(msg_.packet.sourceChannel);
+            _transferFrom(sender, ics20EscrowAddress, erc20Address, packetData.amount);
+        } else {
             // receiver chain is source: burn the vouchers
-            // TODO: Implement escrow balance tracking (#6)
-            IBCERC20 ibcERC20Contract = IBCERC20(erc20Address);
-            ibcERC20Contract.burn(packetData.amount);
+            IBCERC20(erc20Address).burn(sender, packetData.amount);
         }
```

Comment:
When sending tokens: if the chain is the source, then escrow the tokens; otherwise, burn them from the sender's account on the ERC20 contract for those tokens.